crates/evm/traces/src/decoder/mod.rs

```diff
@@ -252,78 +252,102 @@ impl CallTraceDecoder {
 
     /// Decodes all nodes in the specified call trace.
     pub async fn decode(&self, traces: &mut CallTraceArena) {
-        for node in &mut traces.arena {
-            self.decode_function(&mut node.trace).await;
-            for log in node.logs.iter_mut() {
-                self.decode_event(log).await;
-            }
-        }
+        self.decode_functions(traces).await;
+        self.decode_events(traces).await;
     }
 
-    async fn decode_function(&self, trace: &mut CallTrace) {
-        let span = trace_span!("decode_function", label = field::Empty).entered();
+    async fn decode_functions(&self, traces: &mut CallTraceArena) {
+        // Step 1: Collect selectors that need decoding.
```

Comment:
At first sight, there are a lot of changes, but many of them are just indent changes. Try `git diff --ignore-space-change